### ISSUE TITLE
How to use split-gpm

### ISSUE THREAD

**Issue by boiscat** [karltayeb/ipsc_gp_clustering#1]
Hi, 
I have a gene expression set with many samples and want to use  split-gpm to cluster the different gene sets and samples. Sorry for that I‘m a green hand for computing area. Do you mind showing me the format of the input file and orders of running these python scripts?

**Comment by karltayeb** [comment]
The SplitGPM-simulation-example.ipynb  is a good place to start.

Briefly you will need to supply a input matrix X, and output matrix Y, and initializations for assignment matrices W1 and W2 for the two levels of clustering

X will be your input data, along with two extra columns indexing your genes and some other categorical feature of the data-- e.g cell for single cell or cell-line as in the paper.

Y will be you gene expression data.

If you copy the notebook, and replace the simulated data with your data you should have a place to start. If you explain your data in more detail I might be able to provide a more specific answer.

**Comment by boiscat** [comment]
Thanks. I will try them first.

**Comment by boiscat** [comment]
Hi karltayeb,

My data sets are correspond to differentation and like the datas in your ipsc paper. I also studied SplitGPM-simulation-example.ipynb and SplitGPM-ipsc-example.ipynb and have som questions.

 ### **metadata:**

>  time:  0,1,2,3 (day)... 
    sample:  SRR01,SRR02
    apa_data:  range from (0-100) 

 
### **X and Y format**

X:                                             

>  time     sample  gene               
 0  1  0    
 0  1  1                    
 1  2  0 
 1  2  1                            
 2  3  0
 2  3  1

Y:

> apa_data
0
25
50
75
100

( APA value is the  quantification of alternative polyadenylation which ranges from 0 to 100)

### **Question 1**
In SplitGPM-simulation-example.ipynb， I changed `timepoints = np.linspace(0, 1, T)` to `timepoints = np.linspace(0, 10, T)` and source the script with some strange predict plots (not good result I guess). Can I  use the interger time value as example format of X above?

In ipsc-example.ipynb,  the command `Z = np.linspace(0, 15, 16)[:, None].astype(np.float64)` is also the time input. I haven't the raw data of your example so I can't get the time in X file. What's the difference of X and Z?

### **Question 2**

To test APA, I guess some scripts could be changed to keep the value static, such as :

>  y += np.random.normal() * 0.1

or

> normalized_data = (data.values - data.values.mean(1)[:, None]) / data.values.std(1)[:, None]

Would these changes influence split-GPM predict efficacy ?

### **Question 3**
Is that sure if my SplitGPM-simulation-example.ipynb works well , other script can also get the reliable results?

**Comment by karltayeb** [comment]
### Question 1
If you only changed `timepoints` and didn't change `Z` that would be a difficult initialization to start training from. Z are inducing points used by the variational approximation for training the model. You can think of the as a smaller set of inputs where that the model will estimate the function at. GPs are expensive to do inference on-- but if you have hundreds of observations at the same input, you can perfectly summarize it with an inducing point and an estimate of the function at that point. Ihttp://proceedings.mlr.press/v5/titsias09a/titsias09a.pdf that's the paper to read if you want to know more about the variational approximation. But I imagine in your case it should look like
`Z = [[0., 1., 2., 3., ...]]`

But in fact, you shouldn't worry about `timepoints` at all, that is simulated data that just gets put into `X`. You should use whatever the collection times of your data are--but do cast them as floats because tensorflow cares about types.

### Question 2

Those are not relevant, that code is there to prepare `Y` from the data I used. You have your own data, the APA values. So you should set Y to be that. However, you should note that the model assumes Gaussian noise. People model log counts with a Gaussian distribution, I do not know if it is appropriate to do that with APA. You should either transform the data to something that looks normal or pick an appropriate noise distribution.

### Question 3

Either notebook has what you need to get running, but it also has extra code for either generating the simulated data or preparing our gene expression data to be passed into the model. This is essentially what you need.

Set up the data and initialize the cluster assignments with
```
X = # input each row looks like [time, gene_id, sample_id]
Y = # output corresponding to each input in X

W1 = # [N, K1] matrix, rows sum to 1
W2 = # [G, K2] matrix, rows sum to 1

N = # number of samples
G = # number of genes
K1 = # number of sample clusters
K2 = # number of gene clusters
```

Initialize the model with
```
name = 'name_for_your_model'
minibatch_size = 500

with gpflow.defer_build():
    kernel = mk.SharedIndependentMok(gpflow.kernels.RBF(1, active_dims=[0]), K1 * K2)
    Z = # set this to something that makes sense for your problem
    feature = gpflow.features.InducingPoints(Z)
    feature = mf.SharedIndependentMof(feature)

    model = SplitGPM(X, Y,
                   np.log(W1_init + 1e-5), np.log(W2_init + 1e-5),
                   kernel, gpflow.likelihoods.Gaussian(),
                   feat=feature,
                   minibatch_size=minibatch_size,
                   name=name)
model.compile()
```

Then you can specify what you want to learn during training with

```
model.W1.set_trainable(True)  # learn cell assignments
model.W2.set_trainable(True)  # learn gene assignments
model.feature.set_trainable(False)  # move inducing points
model.kern.set_trainable(True)  # learn kernel parameters
model.likelihood.set_trainable(True)  # lear likelihood parameters
```

and train the model with 

```
adam = gpflow.train.AdamOptimizer(0.005)
adam.minimize(model, maxiter={set this number})
```

Hope this helps. Admittedly, I should write a script that takes care of some of these details for people that just want to run the mode. I chose to leave them exposed because setting them up well is important to getting reasonable outputs from the model. I will think on how to make it easier to run and hopefully push a script or something sometime next week. In the meantime I hope this helps.
